web/modules/custom/joinup_migrate/joinup_migrate.post_update.php

```diff
@@ -159,3 +158,86 @@ function joinup_migrate_post_update_more_redirects() {
     }
   }
 }
+
+/**
+ * Fix distribution files with the same name [ISAICP-3968].
+ */
+function joinup_migrate_post_update_same_name_distribution_files() {
+  /** @var \Drupal\Core\File\FileSystemInterface $file_system */
+  $file_system = \Drupal::service('file_system');
+  /** @var \Drupal\Component\Uuid\UuidInterface $uuid_generator */
+  $uuid_generator = Drupal::service('uuid');
+  /** @var \Drupal\Core\Entity\ContentEntityStorageInterface $redirect_storage */
+  $redirect_storage = \Drupal::entityTypeManager()->getStorage('redirect');
+
+  $legacy_db = Database::getConnection('default', 'migrate');
+  $legacy_db_name = $legacy_db->getConnectionOptions()['database'];
+
+  /** @var \Drupal\Core\Database\Query\SelectInterface $query */
+  $query = Database::getConnection()->select('file_managed', 'f')
+    ->fields('f', ['fid', 'uri'])
+    ->fields('d6f', ['filename', 'filepath'])
+    ->orderBy('f.uri')
+    ->condition('f.uri', 'public://distribution/%', 'LIKE');
+  $query->join("$legacy_db_name.files", 'd6f', 'f.fid = d6f.fid');
+  $file_groups = [];
+  foreach ($query->execute()->fetchAll() as $row) {
+    $dir = substr($file_system->dirname($row->uri), 22);
+    // We accept only date formatted dirs.
+    if (preg_match('@^\d{4}\-\d{2}$@', $dir)) {
+      $file_groups[$dir][$row->filename][$row->fid] = $row->filepath;
+    }
+  }
+
+  // Keep only duplicates.
+  array_walk($file_groups, function (array &$file_group) {
+    $file_group = array_filter($file_group, function (array $files) {
+      return count($files) > 1;
+    });
+  });
+  $file_groups = array_filter($file_groups, function (array $file_group) {
+    return (bool) $file_group;
+  });
+
+  $prefix = "public://distribution/access_url";
+  $redirects = [];
+  foreach ($file_groups as $month => $file_group) {
+    foreach ($file_group as $file_name => $files) {
+      foreach ($files as $fid => $source_path) {
+        $destination = "$prefix/$month/{$uuid_generator->generate()}";
+        // Create the destination directory.
+        if (!is_dir($destination)) {
+          $file_system->mkdir($destination, NULL, TRUE);
+        }
+        $file = File::load($fid);
+
+        // Ensure the filename.
+        $file->setFilename($file_name);
+        if (!$file = file_move($file, "$destination/$file_name")) {
+          throw new \Exception("Can't move file $file_name (fid $fid).");
+        }
+        $redirects[$source_path] = "base:/sites/default/files/" . str_replace('public://', '', $file->getFileUri());
+      }
+    }
+  }
+
+  // Delete existing redirects.
+  if ($rids = $redirect_storage->getQuery()
```

Comment:
So we are creating new redirects for the moved files, so the links that are out already are not turned into dead ends. This is really clever! But is it needed the step of deleting existing redirects? As far as I know there should not be any others to files. Just asking.

Reply:
I've created redirects to all migrated files so that we lower the broken links. Such files were migrated, thus the have redirects. Now we want to override those redirects.

Reply:
👍